File: Algorithms/py/099.RecoverBinarySearchTree.py
class Solution(object):
    def recoverTree(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        self.pre, self.node1, self.node2 = None, None, None
        self.recoverTreeRecursive(root)
        self.node1.val, self.node2.val = self.node2.val, self.node1.val

    def recoverTreeRecursive(self, root):
        if not root:
            return
        self.recoverTreeRecursive(root.left)
        if self.pre and root.val < self.pre.val:
            if not self.node1:
                self.node1 = self.pre
                self.node2 = root
            else:
                self.node2 = root
        self.pre = root
        self.recoverTreeRecursive(root.right)

File: Algorithms/java/099.RecoverBinarySearchTree.java
public class Solution {
    TreeNode pre, node1, node2;
    public void recoverTree(TreeNode root) {
        pre = null;
        node1 = null;
        node2 = null;
        recoverTreeRecursive(root);
        int temp = node1.val;
        node1.val = node2.val;
        node2.val = temp;
    }
    private void recoverTreeRecursive(TreeNode root) {
        if (root == null) {
            return;
        }
        recoverTreeRecursive(root.left);
        if (pre != null && root.val < pre.val) {
            if (node1 == null) {
                node1 = pre;
                node2 = root;
            } else {
                node2 = root;
            }
        }
        pre = root;
        recoverTreeRecursive(root.right);
    }
}
